main/lsp/requests/code_action.cc

```diff
@@ -29,8 +34,234 @@ vector<unique_ptr<TextDocumentEdit>> getEdits(const LSPConfiguration &config, co
     }
     return documentEdits;
 }
+
+optional<const ast::MethodDef *> findMethodTree(const ast::ExpressionPtr &tree, const core::SymbolRef method) {
+    if (auto seq = ast::cast_tree<ast::InsSeq>(tree)) {
+        for (auto &subtree : seq->stats) {
+            auto maybeMethod = findMethodTree(subtree, method);
+            if (maybeMethod.has_value()) {
+                return maybeMethod.value();
+            }
+        }
+    } else if (auto klass = ast::cast_tree<ast::ClassDef>(tree)) {
+        for (auto &subtree : klass->rhs) {
+            auto maybeMethod = findMethodTree(subtree, method);
+            if (maybeMethod.has_value()) {
+                return maybeMethod.value();
+            }
+        }
+    } else if (auto methodDef = ast::cast_tree<ast::MethodDef>(tree)) {
+        if (methodDef->symbol == method.asMethodRef()) {
+            return methodDef;
+        }
+    }
+    return nullopt;
+}
+
+unique_ptr<string> copyMethodSource(const core::GlobalState &gs, optional<core::LocOffsets> maybeSigLoc,
```

Comment:
Why make this a `unique_ptr`?

`std::string` will manage its own storage. If youu need to get around the fact that `substr` returns a `string_view`, you should be able to wrap the whole thing in a call to `return string(fref.data(gs)...)`

Reply:
I was not sure about string managing it's own lifetime, so I wrapped it into `unique_ptr`